baby-liminal-extension/build.rs

```diff
@@ -0,0 +1,247 @@
+        let mut buffer = Vec::new();
+        buffer.extend(k.to_le_bytes());
+        buffer.extend(vk.to_bytes(halo2_proofs::SerdeFormat::RawBytesUnchecked));
+        buffer
+    }
+}
+
+/// This module defines the circuit for the benchmarking purposes.
+#[cfg(feature = "runtime-benchmarks")]
+mod circuit {
+    use halo2_proofs::{
+        circuit::{Layouter, Region, Value},
+        halo2curves::bn256::Fr,
+        plonk::{Circuit, ConstraintSystem, Error},
+        standard_plonk::{StandardPlonk, StandardPlonkConfig},
+    };
+
+    pub struct BenchCircuit<const INSTANCES: usize, const ROW_BLOWUP: usize> {
+        /// The roots of the instances (`i`th root is a square root of `i`th instance).
+        roots: [Fr; INSTANCES],
+    }
+
+    impl<const INSTANCES: usize, const ROW_BLOWUP: usize> Default
+        for BenchCircuit<INSTANCES, ROW_BLOWUP>
+    {
+        fn default() -> Self {
+            BenchCircuit {
+                roots: [Fr::zero(); INSTANCES],
+            }
+        }
+    }
+
+    impl<const INSTANCES: usize, const ROW_BLOWUP: usize> BenchCircuit<INSTANCES, ROW_BLOWUP> {
+        /// Create a circuit with the consecutive natural numbers as advices.
+        pub fn natural_numbers() -> Self {
+            let roots: Vec<_> = (0..INSTANCES).map(|i| Fr::from(i as u64)).collect();
+            Self {
+                roots: roots.try_into().unwrap(),
+            }
+        }
+
+        /// Assign the `idx`th root to the `a` and `b` advices and `-1` to the `q_ab` fixed value.
+        ///
+        /// We assume here that this is the first row in the region (offset 0).
+        fn neg_root_square(
+            &self,
+            idx: usize,
+            region: &mut Region<Fr>,
+            config: &StandardPlonkConfig<Fr>,
+        ) -> Result<(), Error> {
+            region.assign_advice(|| "", config.a, 0, || Value::known(self.roots[idx]))?;
+            region.assign_advice(|| "", config.b, 0, || Value::known(self.roots[idx]))?;
+            region.assign_fixed(|| "", config.q_ab, 0, || Value::known(-Fr::one()))?;
+            Ok(())
+        }
+    }
+
+    impl<const INSTANCES: usize, const ROW_BLOWUP: usize> Circuit<Fr>
+        for BenchCircuit<INSTANCES, ROW_BLOWUP>
+    {
+        type Config = <StandardPlonk as Circuit<Fr>>::Config;
+        type FloorPlanner = <StandardPlonk as Circuit<Fr>>::FloorPlanner;
+
+        fn without_witnesses(&self) -> Self {
+            BenchCircuit::default()
+        }
+
+        fn configure(meta: &mut ConstraintSystem<Fr>) -> Self::Config {
+            StandardPlonk::configure(meta)
+        }
+
+        fn synthesize(
+            &self,
+            config: Self::Config,
+            mut layouter: impl Layouter<Fr>,
+        ) -> Result<(), Error> {
+            for instance_idx in 0..INSTANCES {
+                // For every instance, we ensure that the corresponding advice is indeed a square root of it.
+                layouter.assign_region(
+                    || "",
```

Comment:
don't know how useful these names are, but you could call it e.g. "negative square root" here

Reply:
annotations added